File: backend/app/schemas/branding.py
"""Branding DTOs."""

import re
from typing import Optional
from pydantic import BaseModel, field_validator


def _validate_hex(v: str) -> str:
    if not re.match(r"^#[0-9A-Fa-f]{6}$", v):
        raise ValueError("Must be a valid hex color (e.g. #1A1A1A)")
    return v.upper()


class BrandingResponse(BaseModel):
    primary_color: str = "#1A1A1A"
    accent_color: str = "#C5D86D"
    background_color: str = "#F0F0F0"
    institute_name: str = "ICT Institute"
    tagline: str = "Learning Management System"
    logo_url: Optional[str] = None
    favicon_url: Optional[str] = None
    preset_theme: Optional[str] = None
    watermark_enabled: bool = True


class BrandingUpdate(BaseModel):
    primary_color: Optional[str] = None
    accent_color: Optional[str] = None
    background_color: Optional[str] = None
    institute_name: Optional[str] = None
    tagline: Optional[str] = None
    logo_url: Optional[str] = None
    favicon_url: Optional[str] = None
    preset_theme: Optional[str] = None

    @field_validator("primary_color", "accent_color", "background_color", mode="before")
    @classmethod
    def validate_hex_color(cls, v):
        if v is not None:
            return _validate_hex(v)
        return v


# ── Certificate Design DTOs ──────────────────────────────────────────

VALID_BORDER_STYLES = {"classic", "modern", "ornate"}


class CertificateDesignResponse(BaseModel):
    primary_color: str = "#1A1A1A"
    accent_color: str = "#C5D86D"
    institute_name: str = "ICT INSTITUTE"
    website_url: str = "https://ict.net.pk"
    logo_url: Optional[str] = None
    title: str = "CERTIFICATE OF COMPLETION"
    body_line1: str = "This is to certify that"
    body_line2: str = "has successfully completed the course"
    sig1_label: str = "Director"
    sig1_name: str = ""
    sig1_image: Optional[str] = None
    sig2_label: str = "Course Instructor"
    sig2_name: str = ""
    sig2_image: Optional[str] = None
    id_prefix: str = "ICT"
    border_style: str = "classic"


class CertificateDesignUpdate(BaseModel):
    primary_color: Optional[str] = None
    accent_color: Optional[str] = None
    institute_name: Optional[str] = None
    website_url: Optional[str] = None
    logo_url: Optional[str] = None
    title: Optional[str] = None
    body_line1: Optional[str] = None
    body_line2: Optional[str] = None
    sig1_label: Optional[str] = None
    sig1_name: Optional[str] = None
    sig1_image: Optional[str] = None
    sig2_label: Optional[str] = None
    sig2_name: Optional[str] = None
    sig2_image: Optional[str] = None
    id_prefix: Optional[str] = None
    border_style: Optional[str] = None

    @field_validator("primary_color", "accent_color", mode="before")
    @classmethod
    def validate_cert_hex(cls, v):
        if v is not None:
            return _validate_hex(v)
        return v

    @field_validator("border_style", mode="before")
    @classmethod
    def validate_border_style(cls, v):
        if v is not None and v not in VALID_BORDER_STYLES:
            raise ValueError(f"border_style must be one of: {', '.join(sorted(VALID_BORDER_STYLES))}")
        return v

    @field_validator("id_prefix", mode="before")
    @classmethod
    def validate_id_prefix(cls, v):
        if v is not None:
            v = v.strip().upper()
            if not 1 <= len(v) <= 10:
                raise ValueError("id_prefix must be 1-10 characters")
            if not re.match(r"^[A-Z0-9]+$", v):
                raise ValueError("id_prefix must be uppercase alphanumeric only")
            return v
        return v

File: frontend/lib/api/branding.ts
import { apiClient } from './client';

export interface BrandingData {
  primaryColor: string;
  accentColor: string;
  backgroundColor: string;
  instituteName: string;
  tagline: string;
  logoUrl: string | null;
  faviconUrl: string | null;
  presetTheme: string | null;
  watermarkEnabled: boolean;
}

export interface BrandingUpdate {
  primaryColor?: string;
  accentColor?: string;
  backgroundColor?: string;
  instituteName?: string;
  tagline?: string;
  logoUrl?: string | null;
  faviconUrl?: string | null;
  presetTheme?: string | null;
}

export interface PresetThemes {
  [key: string]: { primary: string; accent: string; background: string };
}

// Use relative path so requests go through Next.js rewrites (same as apiClient)
const API_BASE = '/api/v1';

/**
 * Public endpoint — fetches branding without auth.
 * Uses raw fetch (not apiClient) since this is called before auth is available.
 * Accepts optional slug to send X-Institute-Slug header for tenant routing.
 */
export async function getBranding(slug?: string | null): Promise<BrandingData> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5000);
  const headers: Record<string, string> = {};
  if (slug) headers['X-Institute-Slug'] = slug;
  let res: Response;
  try {
    res = await fetch(`${API_BASE}/branding`, { signal: controller.signal, headers });
  } finally {
    clearTimeout(timeout);
  }
  if (!res.ok) throw new Error('Failed to fetch branding');
  const data = await res.json();
  // Manual snake_case to camelCase conversion for this public endpoint
  return {
    primaryColor: data.primary_color ?? '#1A1A1A',
    accentColor: data.accent_color ?? '#C5D86D',
    backgroundColor: data.background_color ?? '#F0F0F0',
    instituteName: data.institute_name ?? 'ICT Institute',
    tagline: data.tagline ?? 'Learning Management System',
    logoUrl: data.logo_url ?? null,
    faviconUrl: data.favicon_url ?? null,
    presetTheme: data.preset_theme ?? null,
    watermarkEnabled: data.watermark_enabled !== false,
  };
}

export async function updateBranding(data: BrandingUpdate): Promise<BrandingData> {
  return apiClient<BrandingData>('/branding', {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export async function uploadLogo(file: File): Promise<{ logoUrl: string }> {
  const formData = new FormData();
  formData.append('file', file);
  return apiClient<{ logoUrl: string }>('/branding/logo-upload', {
    method: 'POST',
    body: formData,
  });
}

export async function getPresetThemes(): Promise<PresetThemes> {
  const { getInstituteSlug } = await import('@/lib/utils/subdomain');
  const slug = getInstituteSlug();
  const headers: Record<string, string> = {};
  if (slug) headers['X-Institute-Slug'] = slug;
  const res = await fetch(`${API_BASE}/branding/preset-themes`, { headers });
  if (!res.ok) throw new Error('Failed to fetch preset themes');
  return res.json();
}

// ── Certificate Design ────────────────────────────────────────────────

export interface CertificateDesign {
  primaryColor: string;
  accentColor: string;
  instituteName: string;
  websiteUrl: string;
  logoUrl: string | null;
  title: string;
  bodyLine1: string;
  bodyLine2: string;
  sig1Label: string;
  sig1Name: string;
  sig1Image: string | null;
  sig2Label: string;
  sig2Name: string;
  sig2Image: string | null;
  idPrefix: string;
  borderStyle: string;
}

export async function getCertificateDesign(): Promise<CertificateDesign> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5000);
  let res: Response;
  try {
    const { getInstituteSlug } = await import('@/lib/utils/subdomain');
    const certSlug = getInstituteSlug();
    const certHeaders: Record<string, string> = {};
    if (certSlug) certHeaders['X-Institute-Slug'] = certSlug;
    res = await fetch(`${API_BASE}/branding/certificate-design`, { signal: controller.signal, headers: certHeaders });
  } finally {
    clearTimeout(timeout);
  }
  if (!res.ok) throw new Error('Failed to fetch certificate design');
  const data = await res.json();
  return {
    primaryColor: data.primary_color ?? '#1A1A1A',
    accentColor: data.accent_color ?? '#C5D86D',
    instituteName: data.institute_name ?? 'ICT INSTITUTE',
    websiteUrl: data.website_url ?? 'https://ict.net.pk',
    logoUrl: data.logo_url ?? null,
    title: data.title ?? 'CERTIFICATE OF COMPLETION',
    bodyLine1: data.body_line1 ?? 'This is to certify that',
    bodyLine2: data.body_line2 ?? 'has successfully completed the course',
    sig1Label: data.sig1_label ?? 'Director',
    sig1Name: data.sig1_name ?? '',
    sig1Image: data.sig1_image ?? null,
    sig2Label: data.sig2_label ?? 'Course Instructor',
    sig2Name: data.sig2_name ?? '',
    sig2Image: data.sig2_image ?? null,
    idPrefix: data.id_prefix ?? 'ICT',
    borderStyle: data.border_style ?? 'classic',
  };
}

export async function updateCertificateDesign(data: Partial<CertificateDesign>): Promise<CertificateDesign> {
  return apiClient<CertificateDesign>('/branding/certificate-design', {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

export async function uploadSignature(file: File, position: 1 | 2): Promise<{ imageUrl: string; position: number }> {
  const formData = new FormData();
  formData.append('file', file);
  return apiClient<{ imageUrl: string; position: number }>(`/branding/signature-upload?position=${position}`, {
    method: 'POST',
    body: formData,
  });
}
